libyul/backends/evm/Dominator.h

```diff
@@ -0,0 +1,271 @@
+
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Dominator analysis of a control flow graph.
+ * The implementation is based on the following paper:
+ * https://www.cs.princeton.edu/courses/archive/spr03/cs423/download/dominators.pdf
+ * See appendix B pg. 139.
+ */
+#pragma once
+
+#include <libyul/backends/evm/ControlFlowGraph.h>
+#include <libsolutil/Visitor.h>
+
+#include <vector>
+#include <map>
+#include <set>
+#include <deque>
+
+namespace solidity::yul
+{
+
+template<typename Vertex, typename ForEachSuccessor>
+class Dominator
```

Comment:
I believe I made all changes suggested. Please let me know if I forgot to improve the readability of something.

Reply:
One more thing, but a bit related: this is a generic algorithm, which does not depend on anything Yul specific. It will also likely be useful outside of `libyul` in the future. `libsolutil` seems like a better place for it.